src/constraint.rs

```diff
@@ -1406,3 +1420,67 @@ impl ConstraintValidator for ScaleConstraint {
         Ok(())
     }
 }
+
+/// Implements Ion Schema's `timestamp_precision` constraint
+/// [scale]: https://amzn.github.io/ion-schema/docs/spec.html#timestamp_precision
+#[derive(Debug, Clone, PartialEq)]
+pub struct TimestampPrecisionConstraint {
+    timestamp_precision_range: Range,
+}
+
+impl TimestampPrecisionConstraint {
+    pub fn new(scale_range: Range) -> Self {
+        Self {
+            timestamp_precision_range: scale_range,
+        }
+    }
+
+    pub fn timestamp_precision(&self) -> &Range {
+        &self.timestamp_precision_range
+    }
+}
+
+impl ConstraintValidator for TimestampPrecisionConstraint {
+    fn validate(&self, value: &OwnedElement, type_store: &TypeStore) -> ValidationResult {
+        // get the precision of given timestamp
+        let timestamp_value = match value.as_timestamp() {
+            Some(timestamp_value) => timestamp_value,
+            _ => {
+                // return Violation if value is not timestamp
+                let error_message = if value.is_null() {
+                    format!("expected a timestamp but found {:?}", value)
+                } else {
+                    format!("expected a timestamp but found {}", value.ion_type())
+                };
+                return Err(Violation::new(
+                    "timestamp_precision",
+                    ViolationCode::TypeMismatched,
+                    &error_message,
+                ));
+            }
+        };
+
+        let value_precision = TimestampPrecisionValue::from_timestamp(&timestamp_value);
+
+        // get isl timestamp precision as a range
+        let precision_range: &Range = self.timestamp_precision();
+
+        // return a Violation if the value didn't follow timestamp precision constraint
+        if !precision_range
+            .contains(&(timestamp_value.to_owned()).into())
+            .unwrap()
+        {
```

Comment:
Could you describe what's happening here? This struck me as a bit redundant:

```rust
//         v-- `timestamp_value` is a `&Timestamp`
    &(timestamp_value.to_owned()) // <-- that we turn into an owned `Timestamp`
//  ^-- before turning it back into a `&Timestamp`
```

Am I misunderstanding? Given the signature of [`Range::contains`](https://doc.rust-lang.org/std/ops/struct.Range.html#method.contains), I'm also not sure why the `unwrap()` is needed at the end.

Reply:
I was using `unwrap()` because contains would return an `IonSchemResult`. These `IonSchemaResult` returns  a single kind of `Error` which checks if the given input value(`OwnedElement`) has the same type as the range type to perform contains. Probably changing `Range` to be `Range<T>` could remove these type checks and also for all constraint the type of the data being passed to `contains` already checked before its passed so its safe to `unwrap` these.

Reply:
> Probably changing Range to be Range<T> could remove these type checks and also for all constraint the type of the data being passed to contains already checked before its passed so its safe to unwrap these.

👍